src/Servers/Kestrel/Core/src/Internal/Http2/Http2OutputProducer.cs

```diff
@@ -455,12 +455,16 @@ private async Task ProcessDataWrites()
                     {
                         var endStream = readResult.IsCompleted;
 
+                        flushResult = await _frameWriter.WriteDataAsync(StreamId, _flowControl, readResult.Buffer, endStream, firstWrite, forceFlush: true);
```

Comment:
This ordering change creates a new race condition where the client can receive the EOS flag before the server decrements its stream counter. That could cause the client to start a new stream and the server to refuse it because they disagree on the active stream count.

It seems like shutdown needs to be based on a different metric that includes draining the response.

Reply:
@Tratcher Good point. A fix could be to move the decrement before the final frame write but after flow control is checked that there is capacity. WDYT?

Reply:
Done.

Moved responsibility for calling decrease active stream count from output producer to frame writer.

Reply:
This is gonna conflict with my PR, can I include this with my fix? I guess I could just merge and reject this change...

Reply:
I'd like to get this in first. There's a chance it is backported to 6.0.